Feature/obspace diagnostics

delippi opened the pull request:
~~Made several updates for observation space diagnostic tool. Two new files were added, "dateutils.py" and "netCDF_obspace_diag.py" for date/time manipulation as for computing the observation space statistics from netCDF diagnostic files. A function for selecting which observations to use from the diag file was also created called "filter_obs". This can filter observations by observation type from the bufr obs table (https://emc.ncep.noaa.gov/mmb/data_processing/prepbufr.doc/table_2.htm) as well as by lat/lon/pressure/error. In addition to these major changes, there were some minor updates to unit conversions the corresponding test_units.py and updated some of the example line_plot scripts (one example script was added to include more options).~~

Edit:
Added a function to help with ensemble DA diagnostic tools being developed in PyGSI. The main change includes the file "dateutils.py" which is used for date manipulation. There are various other small changes to units.py, plots.py, create_plots.py, a new example in examples, and some reorganization of tools.

Commit:
add new file "dateutils.py" for date/time manipulation

Commit:
updated mm to inches and vice versa to use exact calculation and added kgkg to gkg and vice versa.

Commit:
added filter_obs function which creates a bool array used to filter by.
Considers obs filtered by the use flag,max/min pressure, latitude, longitude,
and errors along with filtering by bufr code.

Commit:
Added file "netCDF_obspace_diag.py" which computes observation space diagnostics from netCDF diag files.

Commit:
updated examples/line_plots/line_plot and create a new line_plot example with more options.

Commit:
updated examples/line_plots/line_plot and create a new line_plot example with more options.

Commit:
updated filter_obs function to use "_np" instead of "np"

Commit:
fixed error in line_plot when using set_xticks which caused the ylabels to be latituded.

Commit:
fix logical errors in comparison of error min/max with errorinv.

Commit:
reorder some of the code and change how the unit conversions are done for q.

Commit:
remove some of the unit conversions that I added earlier. Not needed.

Commit:
add logic in case of missing netCDF diag files

Commit:
made two fixes: 1) fixed issue with looping multiple experiments and
2) added parameter "skip_enkf_hours" to skip any times for which to ignore.
for example, rrfs is only run at 18-23Z currently. Use this parameter instead
of hardcoding to skip anything <=18.

Commit:
update test_units.py to match updated mm_to_in calc.

Commit:
update print statement

Commit:
update order of arguments and set skip_enkf_hours=[] as default.

CoryMartin-NOAA commented on src/emcpy/io/netCDF.py:
```diff
@@ -79,3 +92,49 @@ def read_netCDF_var(filename, variable_name, oneD=False, ftime=-1, flevel=-1):
     result = _np.squeeze(var)
     return result
+
+def filter_obs(code,codes,errorinv,lat,lon,pressure,\
```
This should probably be somewhere else but not immediately sure where. Any suggestions @aerorahul or @kevindougherty-noaa ?

delippi replied:
I made a utils directory and move this function there.

delippi committed:
created stats/ and utils/ and move the corresponding stats.py and utils.py into
those directories. Updated a few other parts of the code to make these
changes work.

delippi committed:
moved "filter_obs" function out of this file and into utils/

delippi committed:
renamed netCDF_obspace_diag.py and moved into stats/obspace.py

delippi committed:
add scripts/ and added main driver script to it in initial commit.

delippi committed:
Fixed an issue in filter_obs for parameter "use". Also,
added hemisphere options for filtering obs in filter_obs.

delippi committed:
updated function name.

delippi committed:
renamed and updated main driver script. Added timer function to utils.py.

delippi committed:
fixed pycodestyle issues

delippi committed:
fixed import line to correspond to new location of utils.py and stats.py.

delippi committed:
forgot to add line_plot examples to commit for pycodestyle updates.

delippi committed:
forgot to add line_plot examples to commit for pycodestyle updates.

delippi committed:
forgot to add line_plot examples to commit for pycodestyle updates.

kevindougherty-noaa commented on scripts/ensemble_obspace_diag.py:
```diff
@@ -0,0 +1,320 @@
+date2 = "2022123023"  # will skip 00-18
+skip_enkf_hours = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]  # right now rrfs only runs EnKF at 18-23Z
+
+datapath = "/lfs/h2/emc/da/noscrub/donald.e.lippi/rrfs_mon/diags/"  # rrfs_a_conus/"
```
Should we get a temp file saved locally instead of pointing to here?

delippi replied:
That would be nice, but I'm not certain it would make sense to do so. This script uses the RRFS ensemble diag files and there are 30 members, so for 1 day there would be about 150 files (considering just uv) since the ensemble is only run for a limited period each day. For reference, that would be 356M. I could save a subset of the ensemble though, but it is still going to be more than 1 file.

Let me know if you would still like me to add some data to point to locally.

kevindougherty-noaa commented on scripts/ensemble_obspace_diag.py:
```diff
@@ -0,0 +1,320 @@
+    fig.save_figure("obs_diag_%s.png" % (ob_type))
+
+# Calculate time to run script
+utils.toc(tic, "End. ")
```
Could we also get a `.png` of the output from this script so users can see?

delippi replied:
Yes, I can add this.

kevindougherty-noaa commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,171 @@
+
+
+def dateto_hrs_since_day1CE(curdate, mixedcal=True):
+    """given datetime.datetime instance, compute hours since 1-Jan-0001"""
```
Please update docstring to Google style for `pydoc` documentation purposes. See examples here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html as well as other functions in emcpy.

delippi replied:
I've updated the docstrings. Let me know if there is something I need to do differently. I'm still in the learning process.

kevindougherty-noaa committed:
Upgrading examples to most up to date plotting code (#112)

* ugrades to examples scripts

* added missing line in scatter plot with linear regression

* add colorbar back

* add grid

kevindougherty-noaa reviewed (changes requested):
Thank you for all the additions to this repo! I had a couple small comments, with the main one being that all of your docstring in your `dateutils.py` script need to be updated to Google style so that the automated documentation works smoothly. 

I also have updated all the examples that were out of date and should be merged to `develop` shortly.

aerorahul commented on src/emcpy/stats/obspace.py:
```diff
@@ -0,0 +1,230 @@
+import numpy as _np
```
@CoryMartin-NOAA 
This looks like a very specific use case.  Should this be part of `emcpy`?

CoryMartin-NOAA replied:
I guess this becomes a philosophical question of where the cutoff happens and what constitutes 'specific' vs 'generic'.

I see two approaches:
1) one 'synoptic' repository that contains generic tools and specific, shared applications of said tools
2) a generic tool repository and numerous, app/group/branch specific repositories that use the generic tools

I think both have their pros/cons. 1 runs the risk of being hard to manage, 2 runs the risk of duplicated efforts

CoryMartin-NOAA replied:
@delippi after @aerorahul 's comment and thinking about this over the past couple weeks and now with fresh eyes, I think this functionality probably belongs in PyGSI?

delippi replied:
@CoryMartin-NOAA I've removed this from emcpy and am working on getting it into pygsi. Thanks for the suggestion.

aerorahul commented on src/emcpy/plots/create_plots.py:
```diff
@@ -669,26 +669,26 @@ def _set_ylim(self, ax, ylim):
         ax.set_ylim(**ylim)
 
-    def _set_xticks(self, ax, xticks):
+    def _set_xticks(self, ax, xticks, latlon=False):
```
Not sure why a generic method `_set_xticks` needs this `latlon`
Isn't the purpose of `try` already doing this?

delippi replied:
With the try/except method, you might axes label that look like the following (on the bottom plot). 

![line_plot_options](https://user-images.githubusercontent.com/31075260/211086519-45536f7b-43bf-4823-ab99-066a2206708b.png)

However, I think a better solution, which I just figured out when trying to reproduce the issue, is to make sure you also use set_xticklabels like the following:
```
plot2.set_xticks(x2)
plot2.set_xticklabels([str(item) for item in x2])
```

delippi replied:
Although, I don't think the "except" ever gets invoked.

CoryMartin-NOAA replied:
Yeah I don't think the try, except approach is best here and like the solution of being explicit if it isn't required for the average user to do so...

aerorahul commented on scripts/ensemble_obspace_diag.py:
```diff
@@ -0,0 +1,320 @@
+import emcpy.utils.utils as utils
```
I don't think this belongs in `emcpy`.

CoryMartin-NOAA replied:
I like the idea of there being some sort of 'gallery' though of useful examples for people to see how things are used, the 'examples' directory is a bit more basic whereas this would be more practical.

CoryMartin-NOAA replied:
What about putting this script in to GSI-utils? I have priviliges there to review/merge and it might make more sense there since it is using GSI ncdiags

delippi replied:
@CoryMartin-NOAA I've moved this out of emcpy. I will work on getting it into GSI-utils as you have suggested.

delippi committed:
added doc strings to dateutils.py

delippi committed:
updated the doc strings of functions in utils

delippi committed:
adding an example png.

delippi committed:
fixed an error in line_plot_options

delippi committed:
added subset of ensemble diagnostic files for sample plot.

delippi committed:
Merge branch 'develop' of https://github.com/delippi/emcpy into feature/obspace_diagnostics

delippi committed:
remove trailing whitespace in file.

CoryMartin-NOAA commented:
@delippi I know @kevindougherty-noaa suggested adding test data, but these additions are like 20MB worth of diags, that is too much to carry around in the repository, because everyone would need to download that when it is cloned. Can we remove them?

delippi committed:
remove example GSI diagnostic files

delippi committed:
removing some files to be moved into other repositories.

Commit:
remove filter_obs.py to be moved to pygsi.

delippi committed:
removed filter obs from emcpy and move to pygsi.

delippi committed:
add capability for horizontal span.

delippi committed:
Merge branch 'feature/obspace_diagnostics' of https://github.com/delippi/emcpy into feature/obspace_diagnostics

CoryMartin-NOAA commented on src/emcpy/__init__.py:
```diff
@@ -5,8 +5,8 @@
 import emcpy.calcs
-import emcpy.utils
-import emcpy.stats
+import emcpy.utils.utils
```
I don't particularly like this 'redundant' naming, can we add something to the `__init__` files so that all of the existing stuff stays `import emcpy.utils` but then there is `import emcpy.utils.dateutils` also?

delippi replied:
I think I figured out a fix.

CoryMartin-NOAA commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,226 @@
+    """
+    yyyy, mm, dd, hh = splitdate(analdate)
+    analdate = datetime.datetime(yyyy, mm, dd, hh)
+    verifdate = analdate + fcsthr * datetime.timedelta(hours=1)
```
why not `hours=fcsthr`?

delippi replied:
I updated the calculation to do
verifdate = analdate +  datetime.timedelta(hours=fcsthr)
either way should be correct though.

CoryMartin-NOAA commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,226 @@
+    Returns:
+        yyyymmddhh: (date str) date
+    """
+    return "%0.4i" % (yyyy) + "%0.2i" % (mm) + "%0.2i" % (dd) + "%0.2i" % (hh)
```
should use `f-strings` here.

delippi replied:
I updated this to use f-strings.

CoryMartin-NOAA commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,226 @@
+original author: Jeff Whitaker <jeffrey.s.whitaker@noaa.gov>
+"""
+
+__all__ = [
```
in general a lot of these things seem like they would be better suited creating datetime objects...

delippi replied:
I think from a basic user stand point it is easier to use date strings and let these functions use datetime objects as necessary.

CoryMartin-NOAA commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,226 @@
+original author: Jeff Whitaker <jeffrey.s.whitaker@noaa.gov>
+"""
+
+__all__ = [
```
in general a lot of these things seem like they would be better suited creating datetime objects...

CoryMartin-NOAA commented on src/emcpy/utils/dateutils.py:
```diff
@@ -0,0 +1,226 @@
+    return calendar.monthrange(yyyy, mm)[1]
+
+
+if __name__ == "__main__":
```
why is this needed?

delippi replied:
It would be useful for leap years and just general date calculations which might otherwise require hard coding of number of days in a month, for example.

CoryMartin-NOAA replied:
Sorry I mean the `__main__` part

CoryMartin-NOAA commented on src/emcpy/utils/utils.py:
```diff
@@ -172,3 +173,31 @@ def printcolour(text, colour='red'):
 printcolor = printcolour
+
+
+def tic():
```
these function names are not exactly useful... is there a more formal timer/timing module that would be a better thing to include here? like how `logger` works for logging?

delippi replied:
You're right about the names. These were some functions that I cooked up for myself and thought they might be useful here too. Would you rather that I remove them? I'm not sure what the "formal" method for using a timer is; I would have to do some digging. This isn't an integral part to what I need added. Let me know if I should remove or just update the naming.

CoryMartin-NOAA replied:
I suggest removing and we figure out a more robust/comprehensive timer solution later. Does that work for you?

CoryMartin-NOAA reviewed (commented):
sorry @delippi to keep making you change things, but I have more comments/suggestions/questions

delippi committed:
remove redundant names.

delippi committed:
update to use f-string and improve a date calculation.

Commit:
remove some timer utilties and will find another solution later.

Commit:
clean up previously removed items.

kevindougherty-noaa reviewed (commented):
I think this all looks great. Thank you for adding the datetime applications as I think that is going to be very useful! My only request is to add a quick, basic test script and example script for using the new `horizontalspan` option.

kevindougherty-noaa reviewed (changes requested):
Coming back and looking at this a second time, please make sure all docstring in your `dateutils.py` script is correct. Please use correct punctuation and include a description of what ever function does, even if the name is self explanatory. The docstring is used for when the documentation page is created, so we would like it to be as professional and clean as possible.

Commit:
fix docstrings in dateutils.

delippi committed:
Merge branch 'NOAA-EMC:develop' into feature/obspace_diagnostics

delippi committed:
add basic test script for using HorizontalSpan option.

delippi committed:
fix pycodesytle standards.

delippi committed:
Merge branch 'feature/obspace_diagnostics' of https://github.com/delippi/emcpy into feature/obspace_diagnostics

delippi commented:
@kevindougherty-noaa, let me know if my recent changes are to your satisfaction. I added a function in "test_plots.py" to test and demo the HorizontalSpan option. I also updated the doc strings in dateutils.

kevindougherty-noaa reviewed (approved):
Looks great, thanks for making those changes!
